### PR TITLE
Resample

### PR DESCRIPTION
**Pull request by jcrist**
Based on #369 by @cpcloud. Did some cleanups, and cleared up merge conflicts.

### CODE REVIEW

**Comment by mrocklin** [comment]
@cpcloud do you have any thoughts on this PR?

@jcrist can you state the limitations of this iteration of resample?  From the tests it looks like we raise nice errors when the frequency is not supported.  

I wonder if @sinhrks would find this PR of interest.

I'd like to get this in quickly if possible to avoid it going stale again.

**Comment by jcrist** [comment]
As far as I can tell, this is good to go. Works on everything I've tried it on, and errors nicely on the few cases we don't support.

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -543,6 +548,70 @@ def quantiles(self, *args, **kwargs):
...
+                self.divisions[-1], rule)
+        index = pd.date_range(start=start, end=end, freq=rule)
+        newdivs = [start]
+        rule = pd.datetools.to_offset(rule)
+        for div in self.divisions[1:-1]:
+            # TODO: can we compute this value in O(1) instead of O(log n)?
+            pos = np.searchsorted(index.values, div.asm8, side='right')
+            actual_pos = min(len(index) - 1, pos)
+            newdiv = index[actual_pos]
+            assert newdiv.offset == rule
```
Comment (sinhrks) [inline dask/dataframe/core.py]: Please use ``freq``, offset is being deprecated.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -543,6 +548,70 @@ def quantiles(self, *args, **kwargs):
...
+        """
+        # TODO: Probably a better way to do this
+        start, end = pd.tseries.resample._get_range_edges(self.divisions[0],
+                self.divisions[-1], rule)
+        index = pd.date_range(start=start, end=end, freq=rule)
+        newdivs = [start]
+        rule = pd.datetools.to_offset(rule)
+        for div in self.divisions[1:-1]:
+            # TODO: can we compute this value in O(1) instead of O(log n)?
+            pos = np.searchsorted(index.values, div.asm8, side='right')
```
Comment (sinhrks) [inline dask/dataframe/core.py]: ``rule.rollforward`` may be used to get new divisions. Though offset is not upldated.

```
# Assuming index[3] is division value
s = pd.Series([1, 2, 3, 4, 5], index=pd.date_range('2011-01-01', freq='D', periods=5))

rule = pd.datetools.to_offset('A')
rule.rollforward(s.index[3])                                          
# Timestamp('2011-12-31 00:00:00')

rule = pd.datetools.to_offset('H')
rule.rollforward(s.index[3])
# Timestamp('2011-01-04 00:00:00', offset='D')
```

Comment (cpcloud) [inline dask/dataframe/core.py]: There's a case where this doesn't work, but I can't remember exactly what it was. @jcrist i'd try this out and see if it still fails.

**Comment by jcrist** [comment]
Fixed the comments, this could use a review @cpcloud, @sinhrks .

**Comment by jcrist** [comment]
Woops, just found an issue that we didn't test for. Nevermind, this still isn't good.

**Comment by jcrist** [comment]
Using `rollforward` doesn't work, because it doesn't take into account how binning works for `resample`, and instead just increments a timestamp by an offset. I ended up creating a series from the divisions, and calling resample on that to get a new set of divisions. This also takes the other kwargs into account when determining new divisions. Ping @cpcloud, @sinhrks  for review.

**Comment by mrocklin** [comment]
@cpcloud reviewed this and gave it his +1 in conversation.  

I'm very glad to see this in.

**Comment by cpcloud** [comment]
+1

**Comment by mrocklin** [comment]
This may fail with pandas master.

**Comment by cpcloud** [comment]
I've found this to generally be true with every new release of pandas since around 0.15 or so.

**Comment by cpcloud** [comment]
Not limited to dask, of course.

**Comment by mrocklin** [comment]
To be explicit.  This fails on my machine

**Comment by cpcloud** [comment]
Should we make an issue?

**Comment by mrocklin** [comment]
I think that @jcrist is looking into it now

**Comment by jcrist** [comment]
This is perhaps a pandas issue - this errors now, where before it worked:

```python
import pandas as pd
import numpy as np

index = pd.date_range(start='20120102', periods=24*60*3, freq='T').values
index = index[::2]
s = pd.Series(np.arange(len(index), dtype='f8'), index=pd.Index(index))
expected = s.resample('30T', how='nunique')
```
Traceback:
```python
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    expected = s.resample('30T', how='nunique')
  File "/Users/jcrist/Code/pandas/pandas/core/generic.py", line 3368, in resample
    return sampler.resample(self).__finalize__(self)
  File "/Users/jcrist/Code/pandas/pandas/tseries/resample.py", line 82, in resample
    rs = self._resample_timestamps()
  File "/Users/jcrist/Code/pandas/pandas/tseries/resample.py", line 287, in _resample_timestamps
    result = grouped.aggregate(self._agg_method)
  File "/Users/jcrist/Code/pandas/pandas/core/groupby.py", line 2339, in aggregate
    return getattr(self, func_or_funcs)(*args, **kwargs)
  File "/Users/jcrist/Code/pandas/pandas/core/groupby.py", line 2564, in nunique
    sorter = np.lexsort((val, ids))
ValueError: all keys need to be the same shape
```

**Pull request merged**